### DISCUSSION TITLE
Pagination on Custom MySQL Query : any ideas ?

### DISCUSSION THREAD

**Discussion by poirazis** [Budibase/budibase#7313] [Help] [answered]
Is it possible (maybe in a hackish way) to paginate the results of a custom query from MySQL ?

**Comment by R2bEEaton** [comment]
Hey @poirazis!

What do you mean by this? I have a custom MSSQL query which I use in a data provider and I'm able to paginate it just fine using the built-in paginate option limited to 50 rows per page. I didn't have to change my query at all.
![image](https://user-images.githubusercontent.com/34921506/184951227-68d26750-6f02-4694-b934-b5fae41cf4a0.png)

**Reply by poirazis** [reply]
I have a custom MySql query and when I bind it to the data provider under paginate it says it's for data coming from tables only. !

[IMG_20220816_220922.jpg](https://user-images.githubusercontent.com/80512170/184962358-b250fe43-e203-4760-ad84-f6bf51a78fad.jpg)

**Reply by R2bEEaton** [reply]
> I have a custom MySql query and when I bind it to the data provider under paginate it says it's for data coming from tables only. !
> 
> [IMG_20220816_220922.jpg](https://user-images.githubusercontent.com/80512170/184962358-b250fe43-e203-4760-ad84-f6bf51a78fad.jpg)

Hmm, what happens when you try checking the box anyway?

**Reply by poirazis** [reply]
Nothing, checks and unchecks but no pagination controls appear in the Data Provider 🤷

**Reply by poirazis** [reply]
Hey @melohagan sorry to have to tag you. Any ideas on this ? Is this the expected behavior or should i raise a bug ?

**Reply by melohagan** [reply]
There's a bug already open: https://github.com/Budibase/budibase/issues/3974

**Comment by melohagan** [comment, marked as answer]
Hi @poirazis 

Paginating the results from a custom query in MySQL is fairly straight forward.

### Data
Fetch your table. In my case it's a table of people names A-Z.
Next create a query with a binding for *LIMIT* and *OFFSET*. I'm using a basic `select * from ...` but this should work for more complicated queries as well.
![Screenshot 2022-08-19 at 11 00 07](https://user-images.githubusercontent.com/101575380/185595115-ccbead99-86a8-4149-a7f2-c36b390c7f0b.png)

### Design
Add a Form, Data Provider (Paginate query), Table.
Under the table add a container to hold a back and next button, with a number field in between.
![Screenshot 2022-08-19 at 11 11 13](https://user-images.githubusercontent.com/101575380/185596955-05bd2328-5f64-403c-994d-c0a9f7a0178b.png)
 
Give the page number field a name, and default value of 1. 
Click to cog icon on the data provider, and add a JavaScript binding for the offset. 
![Screenshot 2022-08-19 at 11 13 06](https://user-images.githubusercontent.com/101575380/185597266-d4937aef-3e14-4761-8c29-dc99b7a4a8bd.png)
```js
return ($("Form.Fields.Page")-1) * 5
```
*multiply by your limit, in this case 5*

Add conditional UI to the back button to prevent negative offsets.
![Screenshot 2022-08-19 at 11 24 34](https://user-images.githubusercontent.com/101575380/185599094-95f4ab3d-a140-4a0c-b354-ff53e6aa0ec7.png)
*Note: it acts as a Less than or equal*

**Define actions** on the back button:
![Screenshot 2022-08-19 at 11 22 43](https://user-images.githubusercontent.com/101575380/185598781-e8158672-6929-481a-9f87-eff9c462a13c.png)

**Define actions** on the next button:
![Screenshot 2022-08-19 at 11 25 17](https://user-images.githubusercontent.com/101575380/185599217-4e9baaf4-5016-4a84-afde-2f8a326a0282.png)

Hide the next button if the number of table rows is less than or equal to the limit minus one. You may want to be smarter here and have a second query to "look ahead" instead. 

![Screenshot 2022-08-19 at 11 28 14](https://user-images.githubusercontent.com/101575380/185599721-79a783d6-a82b-4b37-8887-437d3088d115.png)

### Result
![pagination_custom_query](https://user-images.githubusercontent.com/101575380/185600874-23a16f6a-56e9-4766-aca2-c13af3903930.gif)

**Reply by poirazis** [reply]
Thank you so much @melohagan !

**Reply by poirazis** [reply]
Hey @melohagan , i have created the pagination,. Fun to create once, but wouldnt want to have to recreate whenever i want to paginate a query.
If you know, what's prohibiting the Data Provider build in paging capabilities to work with MySQL queries ? Is there a limitation somewhere or is it something that hasnt been implemented yet ?

**Reply by melohagan** [reply]
For SQL queries I believe it simply hasn't been implemented yet.

**Reply by poirazis** [reply]
I will have a look if I can do this, and if i can, i will do a PR.

**Reply by melohagan** [reply]
Thanks @poirazis As an FYI we're thinking of opening up discord in the near future to support contributors. Some time in September; we'll let you know closer to the time!

**Reply by poirazis** [reply]
That would be great ! Can you point me out where I should be looking? Is this features part of the MySql connector or part of core ? Where are the connectors located ?

**Reply by melohagan** [reply]
It's not something I'm too familiar with, but looking through the code I can see this flow:

- DataProvider.svelte *calls fetchData.update*
- fetchData.js *creates a new Fetch of type QueryFetch*
- QueryFetch.js *extends DataFetch.js*
- DataFetch.js *update method calls getInitialData which currently determineFeatureFlags, which looks as follows:*

![Screenshot 2022-08-22 at 14 34 31](https://user-images.githubusercontent.com/101575380/185933867-2898113d-2a97-4fc4-af8c-b92a42860426.png)

So if I was working on this, the next step would be to determine what happens if I simply set *supportsPagination* to true. Presumably there's a reason it's not so simple.

Another thing to look at is **mysql.ts** which extends **sql.ts**

The *read* method in there uses **knex** (you can find that library on GitHub). You can see this method handles pagination, but it only works for "datasource plus" i.e. SQL table datasources (not custom queries).

The custom query path however goes through the **internalQuery** method of **mysql.ts**. 

It would need a bit of a play around for myself as well, but hopefully that's a little bit of a starting point for you!

**Reply by ashik-entym** [reply]
This would be a nice feature to have , I current have a use case where I join 3 tables and display it. There is potential for this select query to return large numbers so pagination would be nice. Is there any way to use mysql views instead for this purpose ?

**Reply by daraul** [reply]
I'm in a situation where I need to add sorting to a table that I've already applied this workaround to -- is my only recourse to get this working to rebuild the entire table by hand?

**Reply by melohagan** [reply]
Hey @daraul would it not be a similar process, but including the `ORDER BY` statement in your query?

**Reply by R2bEEaton** [reply]
Any updates on this? A much-needed feature. Both allowing SQL Views as a workaround or just allowing custom queries to be paginated would go a long way.
